File: Rust/rust_aoc/src/year_2016/d5_door_pass.rs
use md5;

const PART_ONE_HINTS: &'static[i32; 8] = &[1739529, 1910966, 1997199, 2854555, 2963716, 3237361, 4063427, 7777889];

pub fn part_one(lines: &Vec<String>) -> String {
    solve_part_one(lines, Some(PART_ONE_HINTS))
}

fn solve_part_one(lines: &Vec<String>, hints: Option<&[i32; 8]>) -> String {
    let prefix = lines[0].clone();
    match hints {
        Some(hints) => {
            hints.iter().map(|hint| prefix.clone() + &hint.to_string())
            .map(|attempt| format!("{:x}", md5::compute(attempt)))
            .map(|hex| hex.as_bytes()[5] as char)
            .collect()    
        },
        None => {
            let mut index = 0;
            let mut result = String::new();
            while result.len() < 8 {
                let attempt = prefix.clone() + &index.to_string();
                let hex = format!("{:x}", md5::compute(attempt));
                if hex.starts_with("00000") {
                    result.push(hex.as_bytes()[5] as char);
                }
                index += 1;
            }
            result                
        }
    }
}
const PART_TWO_HINTS: &'static[i32; 8] = &[1910966, 1997199, 2963716, 3237361, 7777889, 12850790, 12942170, 25651067];

pub fn part_two(lines: &Vec<String>) -> String {
    solve_part_two(lines, Some(PART_TWO_HINTS))
}

fn solve_part_two(lines: &Vec<String>, hints: Option<&[i32; 8]>) -> String {
    let prefix = lines[0].clone();

    match hints {
        Some(hints) => {
            let mut pairs = hints.map(|hint| prefix.clone() + &hint.to_string())
            .map(|attempt| format!("{:x}", md5::compute(attempt)))
            .map(|hex| (hex.as_bytes()[5] - b'0', hex.as_bytes()[6]))
            .into_iter()
            .collect::<Vec<_>>();
            pairs.sort_by(|(lhs, _), (rhs, _)| lhs.cmp(rhs));
            pairs.iter().map(|(_, c)| *c as char)
            .collect::<String>()
        },
        None => {
            let mut result = String::from("________");
            let mut index = 0;
            let mut found_count = 0;
        
            while found_count < 8 {
                let attempt = prefix.clone() + &index.to_string();
                let hex = format!("{:x}", md5::compute(attempt));
                if hex.starts_with("00000")  && hex.as_bytes()[5] < b'8' {
                    let pos = hex.as_bytes()[5] - b'0';
                    if result.as_bytes()[pos as usize] == b'_' {
                        unsafe {
                            result.as_bytes_mut()[pos as usize] = hex.as_bytes()[6];
                        }
                        found_count += 1;
                    }
                }
                index += 1;
            }
            return result;
        }
    }

}

File: Cpp/AdventOfCode/src/2016/d5_DoorPass.cpp
#include "2016/d5_DoorPass.h"
#include "Algorithms/md5Constexpr.h"

SOLUTION(2016, 5) {

	constexpr std::string PartOneImpl(std::string_view key, const std::vector<u64>&hints) {
		std::string prefix = std::string(key);
		if (hints.empty()) {
			u64 index = 0;
			std::string result = "";
			while (result.size() < 8) {
				std::string attempt = prefix + ToString(index);
				auto hex = ToHexLower(md5::compute(attempt.c_str()));
				if (hex.substr(0, 5) == "00000") {
					result += hex[5];
				}
				index++;
			}
			return result;
		}
		else {
			std::string result = "";
			for (auto hint : hints) {
				std::string attempt = prefix + ToString(hint);
				auto hex = ToHexLower(md5::compute(attempt.c_str()));
				if (hex.substr(0, 5) != "00000") {
					throw "Bad hint";
				}
				result += hex[5];
			}

			return result;
		}
	}

	PART_ONE() {
		std::vector<u64> hints = {
			1739529,
			1910966,
			1997199,
			2854555,
			2963716,
			3237361,
			4063427,
			7777889
		};
		return PartOneImpl(lines[0], hints);
	}

	constexpr std::string PartTwoImpl(std::string_view key, const std::vector<u64>& hints) {
		std::string result = "________";
		std::string prefix = std::string(key);

		if (hints.empty()) {
			u64 index = 0;

			int totalFound = 0;
			while (totalFound < 8) {
				std::string attempt = prefix + ToString(index);
				auto hex = ToHexLower(md5::compute(attempt.c_str()));
				if (hex.substr(0, 5) == "00000" && hex[5] < '8') {
					auto pos = hex[5] - '0';
					if (result[pos] == '_') {
						result[pos] = hex[6];
						totalFound++;
					}
				}
				index++;
			}
			
		}
		else {
			for (const auto hint : hints) {
				std::string attempt = prefix + ToString(hint);
				auto hex = ToHexLower(md5::compute(attempt.c_str()));
				if (hex.substr(0, 5) != "00000" || hex[5] >= '8') throw "Bad hint";

				auto pos = hex[5] - '0';
				if (result[pos] != '_') throw "Bad Hint";
				result[pos] = hex[6];
			}
		}

		return result;
	}

	PART_TWO() {
		std::vector<u64> hints = {
			1910966,
			1997199,
			2963716,
			3237361,
			7777889,
			12850790,
			12942170,
			25651067
		};
		return PartTwoImpl(lines[0], hints);
	}

	TESTS() {
		std::vector<u64> hints = {
			3231929,
			5017308,
			5278568,
			5357525,
			5708769,
			6082117,
			8036669,
			8605828
		};

		if (PartOneImpl("abc", hints) != "18f47a30") return false;
		return true;
	}
}